rust/et-embedding-worker/src/lib.rs

```diff
@@ -11,10 +13,121 @@ pub mod config;
 pub mod metric_consts;
 
 pub async fn handle_batch(
-    _fingerprints: Vec<NewFingerprintEvent>,
+    fingerprints: Vec<NewFingerprintEvent>,
     _offsets: &[Offset],
-    _context: Arc<AppContext>,
+    context: Arc<AppContext>,
 ) -> Result<Vec<EmbeddingRecord>> {
-    // Your implementation here
-    Ok(vec![]) // For now, we'll just eat the new fingerprint messages
+    let mut handles = vec![];
+    let client = Client::new();
+    let api_key = context.config.openai_api_key.clone();
+    for fingerprint in fingerprints {
+        handles.push(generate_embedding(
+            client.clone(),
+            api_key.clone(),
+            fingerprint,
+        ));
+    }
+    let results = futures::future::join_all(handles).await;
+    results.into_iter().collect()
+}
+
+const OPENAI_EMBEDDING_MODEL: &str = "text-embedding-3-small";
+const EMBEDDING_VERSION: i64 = 1;
+
+#[derive(Serialize)]
+struct OpenAIEmbeddingRequest {
+    input: String,
+    model: String,
+}
+
+#[derive(Deserialize)]
+struct OpenAIEmbeddingResponse {
+    data: Vec<EmbeddingData>,
+}
+
+#[derive(Deserialize)]
+struct EmbeddingData {
+    embedding: Vec<f64>,
+}
+
+pub async fn generate_embedding(
+    client: Client,
+    api_key: String,
+    fingerprint: NewFingerprintEvent,
+) -> Result<EmbeddingRecord> {
+    // Generate text representation of the exception and frames
+    let text = generate_text_representation(&fingerprint);
+
+    // Call OpenAI API to generate embeddings
+    let request = OpenAIEmbeddingRequest {
+        input: text,
+        model: OPENAI_EMBEDDING_MODEL.to_string(),
+    };
+
+    let response = client
+        .post("https://api.openai.com/v1/embeddings")
+        .header("Authorization", format!("Bearer {api_key}"))
+        .header("Content-Type", "application/json")
+        .json(&request)
+        .send()
+        .await?;
+
+    let response_body: OpenAIEmbeddingResponse = response.json().await?;
+
+    let embeddings = response_body
+        .data
+        .first()
+        .ok_or_else(|| anyhow::anyhow!("No embedding data returned from OpenAI"))?
+        .embedding
+        .clone();
+
+    Ok(EmbeddingRecord::new(
+        fingerprint.team_id,
+        OPENAI_EMBEDDING_MODEL.to_string(),
+        EMBEDDING_VERSION,
+        fingerprint.fingerprint.to_string(),
+        embeddings,
+    ))
+}
+
+fn generate_text_representation(fingerprint: &NewFingerprintEvent) -> String {
+    let mut text_parts = Vec::new();
+
+    for exception in &fingerprint.exception_list {
+        // Add exception type and value
+        text_parts.push(format!(
+            "{}: {}",
+            exception.exception_type, exception.exception_value
+        ));
+
+        // Add frame information
+        for frame in &exception.frames {
```

Comment:
i think we should skip all non in app frames here, eg exclude all 3rd party frames, OS frames, browser frames, lang frames etc and just keep the bare min which is the code that the user actually has written